### ISSUE TITLE
Tpetra::CrsMatrix: Add internal thread-safe version of getCrsGraph

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#1336]
@trilinos/tpetra 
Blocks: #941 

`CrsMatrix::getCrsGraph` returns a `Teuchos::RCP` of
the matrix's graph.  This is not currently thread safe, because RCP's
reference count is not currently thread safe.  In particular, RCP's
copy constructor, assignment operator (`operator=`), and destructor all
modify the reference count, so none of these are thread safe.  The
`getCrsGraph` method must increment the graph's reference count, since
it returns the RCP by value.  Thus, it cannot be thread safe.
Returning `const RCP&` is not safe (threads or no), because callers
can access it even after the lifetime of the graph has ended.

Today, I was reminded of this lack of thread safety the hard way,
while host-thread-parallelizing `CrsMatrix::mergeRedundantValues` (see
#941).  Many methods in CrsMatrix look "harmless" -- they appear just
to access some local state like the number of rows owned by the
calling process -- but are not thread safe.  This is because they call
`getCrsGraph`, then ask the graph for information.  Many of the
methods that ask the graph for information are themselves thread safe,
but the `getCrsGraph` call itself is the problem.  I was able to fix
the thread safety issue in my changes to `mergeRedundantValues`, by
lifting the relevant call (`isStorageOptimized()`) outside of the
host-thread-parallel loop.  However, I would like to forestall this
issue for future work.

I propose the following fix: add a new private method, `getCrsGraphRef`, to CrsMatrix.
The new method will return `const crs_graph_type&`.  Its implementation must
carefully avoid any RCP methods that would change the RCP's reference
count.

**Comment by jwillenbring** [comment]
The proposed fix for RCP thread safety is currently under performance evaluation. This is likely to be applied soon, with the default off. I will let others ( @krcb @bartlettroscoe @MicheldeMessieres ) comment further.

**Comment by mhoemmen** [comment]
@jwillenbring Thanks for clarifying!  The above changes would still help performance, since updating the reference count in a thread-safe way would require an atomic update.  Many concurrent atomic updates to the same memory location won't perform well.  Thus, it still makes sense to avoid RCP (or std::shared_ptr, or Kokkos::View) reference count updates if possible.

**Comment by bartlettroscoe** [comment]
> Many concurrent atomic updates to the same memory location won't perform well. Thus, it still makes sense to avoid RCP (or std::shared_ptr, or Kokkos::View) reference count updates if possible.

Reference count updates in non-debug-mode code can be avoided by using iterators, Teuchos::ArrayView and Teuchos::Ptr instead of directly using Teuchos::ArrayPtr and Teuchos::RCP.  See the section "5.12 Optimized performance" in:

* https://trilinos.org/docs/dev/packages/teuchos/doc/html/TeuchosMemoryManagementSAND.pdf

If you are careful about how you organize your code, you can get near optimal runtime performance but still catch programming errors in debug-mode builds.

Making the Teuchos Memory Management classes thread-safe will allow it to be used in threaded code and catch many programming errors.

**Comment by mhoemmen** [comment]
Thanks @bartlettroscoe !  The issue here is not related to iterators.  It's about how the CrsMatrix accesses its CrsGraph.  Imagine some user writes the following code, for a CrsMatrix A:
```
for (LocalOrdinal i = 0; i < A.getNodeNumRows(); ++i) { 
  doSomething(A, i);
}
```
`CrsMatrix::getNodeNumRows` was implemented by asking the matrix's CrsGraph:
```
LocalOrdinal
CrsMatrix<...>::getNodeNumRows() {
  return this->getCrsGraph()->getNodeNumRows();
}
```
The method `getCrsGraph` returns RCP, by value.  We don't want to return RCP by reference; that's not safe (for reasons not related to threads).  Returning RCP by value increments the reference count.  The compiler doesn't know that the loop bound won't change, so it has to evaluate the loop bound on every iteration.  This means every iteration of the loop will touch the reference count.

I avoided this issue by adding a new private method, getCrsGraphRef.  I made it private because it is not safe (for reasons not related to threads) for users to have a reference to the graph, but it is safe in the scope of a CrsMatrix method.

**Comment by bartlettroscoe** [comment]
But CrsMatrix could return a semi-peristing view of this object and return it as a Teuhos::Ptr object.  That will not touch a reference count in a non-debug-mode build.

**Comment by mhoemmen** [comment]
@bartlettroscoe it's ok, don't worry :)  this is just an implementation detail.  also, reference is the right idiom, not pointer, since the graph always needs to exist.

**Comment by bartlettroscoe** [comment]
>  reference is the right idiom, not pointer, since the graph always needs to exist.

the issue is that `Teuchos::Ptr<T>` will catch dangling references but `T&` will not.  That is why you almost always want to sue `Teuchos::Ptr<T>` for sharing objects where a dangling references is a danger and the overhead is small.

**Comment by mhoemmen** [comment]
Dangling references are not a danger in this case.  Also, I need thread safety in debug builds.  Sorry to be so insistent :)
